app/src/main/java/org/thoughtcrime/securesms/conversation/v2/mention/MentionViewModel.kt

```diff
@@ -0,0 +1,272 @@
+import android.text.Spanned
+import android.text.style.StyleSpan
+import androidx.core.text.getSpans
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedInject
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.withContext
+import org.session.libsession.messaging.contacts.Contact
+import org.thoughtcrime.securesms.database.DatabaseContentProviders.Conversation
+import org.thoughtcrime.securesms.database.GroupDatabase
+import org.thoughtcrime.securesms.database.GroupMemberDatabase
+import org.thoughtcrime.securesms.database.MmsDatabase
+import org.thoughtcrime.securesms.database.SessionContactDatabase
+import org.thoughtcrime.securesms.database.Storage
+import org.thoughtcrime.securesms.database.ThreadDatabase
+import org.thoughtcrime.securesms.util.observeChanges
+
+/**
+ * A ViewModel that provides the mention search functionality for a text input.
+ *
+ * To use this ViewModel, you (a view) will need to:
+ * 1. Observe the [autoCompleteState] to get the mention search results.
+ * 2. Set the EditText's editable factory to [editableFactory], via [android.widget.EditText.setEditableFactory]
+ */
+class MentionViewModel(
+    threadID: Long,
+    contentResolver: ContentResolver,
+    threadDatabase: ThreadDatabase,
+    groupDatabase: GroupDatabase,
+    mmsDatabase: MmsDatabase,
+    contactDatabase: SessionContactDatabase,
+    memberDatabase: GroupMemberDatabase,
+    storage: Storage,
+) : ViewModel() {
+    private val editable = MentionEditable()
+
+    /**
+     * A factory that creates a new [Editable] instance that is backed by the same source of truth
+     * used by this viewModel.
+     */
+    val editableFactory = object : Editable.Factory() {
+        override fun newEditable(source: CharSequence?): Editable {
+            if (source === editable) {
+                return source
+            }
+
+            if (source != null) {
+                editable.replace(0, editable.length, source)
+            }
+
+            return editable
+        }
+    }
+
+    @Suppress("OPT_IN_USAGE")
+    private val members: StateFlow<List<Member>?> =
+        (contentResolver.observeChanges(Conversation.getUriForThread(threadID)) as Flow<Any?>)
+            .debounce(500L)
+            .onStart { emit(Unit) }
+            .mapLatest {
+                val r = checkNotNull(threadDatabase.getRecipientForThreadId(threadID)) {
```

Comment:
Not a huge deal, but would be good to avoid one letter variables, just for readability, would be good to use meaningful names.

Reply:
Good call-out. Updated.